### PR TITLE
Add workaround to fix include warnings in Python 2 builds.

### PR DESCRIPTION
**Pull request by zdevito**
This prevents warnings like:

```
[1/1] gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall ...H_CUDNN -MMD -MF build/temp.linux-x86_64-2.7/torch/csrc/jit/script/compiler.d
In file included from /home/zdevito/local/anaconda2/include/python2.7/Python.h:8:0,
                 from /data/users/zdevito/pytorch/torch/csrc/utils/pybind.h:3,
                 from /data/users/zdevito/pytorch/torch/csrc/jit/tracer.h:12,
                 from /data/users/zdevito/pytorch/torch/csrc/autograd/function.h:9,
                 from /data/users/zdevito/pytorch/torch/csrc/jit/generated/aten_dispatch.h:3,
                 from torch/csrc/jit/script/compiler.cpp:3:
/home/zdevito/local/anaconda2/include/python2.7/pyconfig.h:1190:0: warning: "_POSIX_C_SOURCE" redefined [enabled by default]
```
for compilation units that do not include Python.h first.

Using CI to see if this works with all of our compilers.

### CODE REVIEW

**Comment by ezyang** [comment]
CI has miraculously passed, but it still makes me nervous because you just *know* someone's going to pipe up and be like "yo `push_macro` doesn't work on my compiler."

Could we... maybe do a feature test for it, in cmake?

**Comment by goldsborough** [comment]
I'm currently fighting the same problem. I think this will mess with the non-python builds, e.g. the C++ API?

**Comment by goldsborough** [comment]
I guess we can wrap it with #ifndef NO_PYTHON

**Comment by zdevito** [comment]
It shouldn't affect the C++ builds because they don't include python_headers.h

push_macro is a pragma, if the compiler doesn't know it, it will be ignored.

In this case, we just let Python.h set the _XOPEN_SOURCE thing.

So for user to see a failure, we'd need:
1. the compiler doesn't support push_macro
2. somehow the undef/redef of _XOPEN_SOURCE actually matters.

That seems pretty unlikely to me.

**Comment by zdevito** [comment]
Can we merge this and see if it works widely? My previous patches to move Python.h to the top of compilation units don't get merged either, and sorting through these warnings is difficult.

**Comment by zdevito** [comment]
(we should wait until the release branch is cut though)

**Review by ezyang — APPROVED** [review]

**Comment by zdevito** [comment]
Let's hope this works :)

**Pull request merged**
